enhancements/baremetal/an-slo-for-baremetal.md

```diff
@@ -0,0 +1,432 @@
+   (as per openshift/api#124)
+3. Expose a `/metrics` endpoint for Prometheus to be configured to
+   scrape (via a `ServiceMontitor`) and define any relevant Prometheus
+   alert rules based on those metrics.
+
+### "Not In Use" SLO Behaviors
+
+Unlike most other SLOs, CBO is not applicable to all cluster
+configurations. On clusters running on an infrastructure platform
+other than `baremetal` it should adhere to the emerging expected
+behaviors for "not in use" SLOs, including:
+
+1. Setting its `ClusterOperator` condition to `Disabled=true`,
+  `Available=true`, `Progressing=False` with appropriate messages.
+2. User interfaces should convey this disabled state differently than
+   a failure mode (e.g. by graying it out).
+
+Currently, insights-operator is the only other example of an SLO
+following this pattern of using a `Disabled` status. Other somewhat
+similar cases following different patterns include:
+
+- Image registry and samples have a `Removed` management state where
+  `Degraded=False`, `Progressing=False`, `Available=True` with
+  `Reason=currentlyUnmanaged`.
+- The cluster credentials operator has a `disabled` config map setting
+  that can be used to disable the operator and it then sets
+  ClusterOperator status conditions to `Degraded=False`,
+  `Progressing=False`, `Available=True` with
+  `Reason=OperatorDisabledByAdmin` for all three conditions.
+
+### cluster-baremetal-operator Details
+
+The CBO will:
+
+- Be a new `openshift/cluster-baremetal-operator` project.
+- Publish an image called `cluster-baremetal-operator`.
+- Use CVO run-level 30, so its manifests will be applied in parallel
+  with the MAO.
+- Add a new `baremetal` `ClusterOperator` with an additional
+  `Disabled` status for non-baremetal platforms.
+- Use the existing `openshift-machine-api` namespace where the
+  BareMetalHost resources are also located.
+- Install the `metal3.io` `Provisioning` (cluster-scoped) and
+  `BareMetalHost` (namespaced) CRDs.
+- Run under a new `openshift-machine-api/cluster-baremetal-operator`
+  `ServiceAccount`.
+- Be launched by a `openshift-machine-api/cluster-baremetal-operator`
+  `Deployment`, copying much of the MAO pod spec in terms of
+  `system-node-critical` priority class, running on masters, security
+  context, resource requests, etc.
+- Implement a controller reconciling the singleton
+  `provisioning-configuration` cluster-scoped `Provisioning` resource
+- Do nothing except set `Disabled=true`, `Available=true`, and
+  `Progressing=False` when the `Infrastructure` resource a platform
+  type other than `BareMetal`.
+- Based on the values in the `Provisioning` resource, create a
+  `metal3` `Deployment` and associated `metal3-mariadb-password` under
+  the `openshift-machine-api` namespace. This is the same as the MAO
+  currently creates.
+
+### Test Plan
+
+### Graduation Criteria
+
+### Upgrade / Downgrade / Verson Skew Strategy
+
+The key thing to consider for upgrades is how to smoothly transition
```

Comment:
Every time we restart the pod managed by the deployment we lose the contents of the ironic database, because there is no persistent storage for that data. Any work ironic has in progress would also be lost. So it's better to minimize restarts as much as possible, and one way to do that would be to just take over the existing deployment instead of launching a new one.

Reply:
As noted in another comment - yes, the CBO provides a dedicated controller for the provisioning singleton.

I do like you KISS thinking, but I think the practical realities of this deployment is the main reason we should incur these additional complexities.

Reply:
Added Doug's explanation at the start of this section